onnxruntime/test/shared_lib/test_inference.cc

```diff
@@ -3219,3 +3219,45 @@ TEST(MultiKernelSingleSchemaTest, DuplicateKernel) {
 }
 
 #endif
+
+const static std::thread::id caller_tid = std::this_thread::get_id();
+static std::atomic_bool atomic_wait{false};
+
+TEST(CApiTest, RunAsync) {
+  Ort::SessionOptions session_options;
+  session_options.SetIntraOpNumThreads(2);
+  Ort::Session session(*ort_env, MODEL_URI, session_options);
+
+  const char* input_names[] = {"X"};
+  float x_value[] = {1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f};
+  int64_t x_dim[] = {3, 2};
+
+  auto memory_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
+
+  Ort::Value input_tensors[1] = {
+      Ort::Value::CreateTensor<float>(memory_info, x_value, 6, x_dim, 2),
+  };
+
+  const char* output_names[] = {"Y"};
+
+  Ort::RunOptions run_options;
+
+  Ort::detail::RunAsyncCallbackStdFn callback = [&](std::vector<Ort::Value>& outputs, Ort::Status status) {
+    auto callee_tid = std::this_thread::get_id();
+    EXPECT_NE(caller_tid, callee_tid);
+    EXPECT_TRUE(status.IsOK());
+    EXPECT_EQ(outputs.size(), 1UL);
+    EXPECT_EQ(outputs[0].At<float>({1, 0}), 9.f);
+    atomic_wait.store(true);
+  };
+
+  EXPECT_NO_THROW(session.RunAsync(run_options, input_names, input_tensors, 1, output_names, 1, callback));
+
+  std::chrono::duration<double, std::milli> dur{100};
+  // timeout in about 10 secs
+  for (int i = 0; i < 100 && !atomic_wait.load(); ++i) {
+    std::this_thread::sleep_for(dur);
+  }
+
+  EXPECT_EQ(atomic_wait.load(), true);
+}
```

Comment:
We can probably do better with showing the usage of this feature. Can we use condition variables instead?

Reply:
a condition variable (which must be used along with a mutex) is a heavy-weight utility incurring os kernel access overheads, plus it has spurious calls which must be explicitly addressed.